### PR TITLE
bpo-44032: Defer clearing last thread state until last GC has been run.

### PR DESCRIPTION
**Pull request by markshannon**
https://bugs.python.org/issue44032

### CODE REVIEW

**Comment by markshannon** [comment]
@erlend-aasland Can you confirm that this fixes the issue?

File: Python/pystate.c [hunk 1]
```diff
@@ -324,6 +326,7 @@ interpreter_clear(PyInterpreterState *interp, PyThreadState *tstate)
 
     /* Last garbage collection on this interpreter */
     _PyGC_CollectNoFail(tstate);
+    PyThreadState_Clear(tstate);
```
Comment (vstinner) [inline Python/pystate.c]: For safety, I would prefer to ensure that the tstate belongs to the interpreter if PyInterpreterState_Clear() is called from a different interpreter (even if I don't know if it's currently possible):

```suggestion
    // Don't clear tstate if it belongs to another interpreter
    if (tstate->interp == interp) {
        PyThreadState_Clear(tstate);
    }
```

**Review by vstinner — COMMENTED** [review]
A thread state should remain more or less usable after PyInterpreterState_Clear() is called. I don't think that this PR fix the root issue of https://bugs.python.org/issue44032#msg394104

IMO it would be safer to only release datastack_chunk memory in `tstate_delete_common()`, rather than in `PyThreadState_Clear()`.

This change remains interesting ;-)

**Comment by erlend-aasland** [comment]
> @erlend-aasland Can you confirm that this fixes the issue?

I'll do as soon as I get back on my computer.

**Comment by markshannon** [comment]
@vstinner I don't see how a thread can remain usable after the interpreter has been cleared.

**Comment by erlend-aasland** [comment]
Applying this patch works, though: https://bugs.python.org/file50058/fix.diff

**Comment by erlend-aasland** [comment]
I just sync'ed with `upstream/main` to get @vstinner's fix for issue 44184 (GH-26274, 615069eb08494d089bf24e43547fbc482ed699b8). With that in `main` _and_ this PR, the issue is fixed; thanks! You can disregard my previous comment.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
> A thread state should remain more or less usable after PyInterpreterState_Clear() is called. I don't think that this PR fix the root issue of https://bugs.python.org/issue44032#msg394104
> 
> IMO it would be safer to only release datastack_chunk memory in `tstate_delete_common()`, rather than in `PyThreadState_Clear()`.

FYI, freeing datastack chunks in `tstate_delete_common()` iso. deferring `PyThreadState_Clear` works fine with the reproducer.

File: Python/pystate.c [hunk 2]
```diff
@@ -900,13 +897,6 @@ PyThreadState_Clear(PyThreadState *tstate)
     if (tstate->on_delete != NULL) {
         tstate->on_delete(tstate->on_delete_data);
```
Comment (erlend-aasland) [inline Python/pystate.c:898]: Should this also be moved to `tstate_delete_common`, or is it unrelated to deleting the data stack?

Comment (markshannon) [reply Python/pystate.c:898]: I have no idea what it does, so I'm not touching it :slightly_smiling_face:

Comment (erlend-aasland) [reply Python/pystate.c:898]: Me neither :) I tried moving it, and it seems to work fine either place. BTW, I found a clue in Include:
https://github.com/python/cpython/blob/a9e43615c2e1fc5dd60063c1509e8b1c5daad095/Include/cpython/pystate.h#L134-L143

Comment (erlend-aasland) [reply Python/pystate.c:898]: Also: https://docs.python.org/3/c-api/init.html#c.PyThreadState_Clear:

_"Changed in version 3.9: This function now calls the PyThreadState.on_delete callback. Previously, that happened in PyThreadState_Delete()."_

Comment (erlend-aasland) [reply Python/pystate.c:898]: ... and here:

https://github.com/python/cpython/blob/a9e43615c2e1fc5dd60063c1509e8b1c5daad095/Modules/_threadmodule.c#L1280-L1329

Comment (erlend-aasland) [reply Python/pystate.c:898]: It's added by @vstinner in GH-18296 (issue 39511). I guess he knows if it should stay or follow, then :)

**Review by vstinner — APPROVED** [review]
LGTM.

Calling _PyObject_VirtualFree() in tstate_delete_common() sounds like the good thing to do. I doesn't mean that a tstate must remain usable after calling PyThreadState_Clear(state), it's just about consistency.

**Comment by vstinner** [comment]
> buildbot/AMD64 RHEL7 Refleaks PR — Build done.
> buildbot/AMD64 RHEL8 Refleaks PR — Build done. 
> buildbot/s390x Fedora Refleaks PR — Build done. 

test_asyncio hangs, unrelated error: https://bugs.python.org/issue44112

**Comment by markshannon** [comment]
I'm not worried about asyncio, but test_multiprocessing also fails for one and test_posix for another.

There doesn't seem to a coherent model of how threads and interpreters are related, or what memory belongs to the interpreter or thread.
As a result, cleaning up after a fork is impossible to get right.

**Comment by erlend-aasland** [comment]
> There doesn't seem to a coherent model of how threads and interpreters are related, or what memory belongs to the interpreter or thread.
> As a result, cleaning up after a fork is impossible to get right.

Sounds like an issue should be opened for this, if there's not already one on bpo.

**Comment by vstinner** [comment]
AMD64 RHEL8 Refleaks PR:

```
test_posix failed
...
test_posix leaked [1, 0, 0] file descriptors, sum=1
```

I can be a random error which doesn't fail with re-run in verbose mode, we don't know since the job hangs before test_posix can be re-run.

**Comment by vstinner** [comment]
> I'm not worried about asyncio, but test_multiprocessing also fails for one and test_posix for another.

I fail to see the test_multiprocessing error. multiprocessing tests are full of race conditions. Some are legit bugs, some are bugs in the tests. I fixed tons of bugs in multiprocessing and in its test suite last years, but if you search for "multiprocessing" in the bug tracker, you will see a long list of issues open for several weeks/months.

IMO it's perfectly safe to merge this PR. All issues that you saw are very likely known (not regression of your change).

Clearing the current thread state would be another can of worm, but your PR no longer does that ;-)

**Comment by markshannon** [comment]
Ok, I'll merge this then.
It seems to not make things worse, and probably makes them better.

**Comment by vstinner** [comment]
Yeah, it's fine, in the really worst case, we can just revert the change to revisit the options. But again, IMO this function is perfectly safe.

**Pull request merged**